File: Tests/Tests/STPPaymentMethodGrabPayParamsTest.m
//
//  STPPaymentMethodGrabPayParamsTest.m
//  StripeiOS Tests
//
//  Created by Yuki Tokuhiro on 7/21/20.
//  Copyright © 2020 Stripe, Inc. All rights reserved.
//

#import <XCTest/XCTest.h>

#import "STPTestingAPIClient.h"

@interface STPPaymentMethodGrabPayParamsTest : XCTestCase

@end

@implementation STPPaymentMethodGrabPayParamsTest


- (void)testCreateGrabPayPaymentMethod {
    STPAPIClient *client = [[STPAPIClient alloc] initWithPublishableKey:STPTestingSGPublishableKey];
    STPPaymentMethodGrabPayParams *grabPayParams = [STPPaymentMethodGrabPayParams new];
    
    STPPaymentMethodBillingDetails *billingDetails = [STPPaymentMethodBillingDetails new];
    billingDetails.name = @"Jenny Rosen";
    
    STPPaymentMethodParams *params = [STPPaymentMethodParams paramsWithGrabPay:grabPayParams
                                                                billingDetails:billingDetails
                                                                      metadata:@{@"test_key": @"test_value"}];
    
    XCTestExpectation *expectation = [self expectationWithDescription:@"Payment Method GrabPay create"];
    
    [client createPaymentMethodWithParams:params
                               completion:^(STPPaymentMethod * _Nullable paymentMethod, NSError * _Nullable error) {
        [expectation fulfill];
        
        XCTAssertNil(error, @"Unexpected error creating GrabPay PaymentMethod: %@", error);
        XCTAssertNotNil(paymentMethod, @"Failed to create GrabPay PaymentMethod");
        XCTAssertNotNil(paymentMethod.stripeId, @"Missing stripeId");
        XCTAssertNotNil(paymentMethod.created, @"Missing created");
        XCTAssertFalse(paymentMethod.liveMode, @"Incorrect livemode");
        XCTAssertEqual(paymentMethod.type, STPPaymentMethodTypeGrabPay, @"Incorrect PaymentMethod type");
        
        // Billing Details
        XCTAssertEqualObjects(paymentMethod.billingDetails.name, @"Jenny Rosen", @"Incorrect name");
        
        // GrabPay Details
        XCTAssertNotNil(paymentMethod.grabPay, @"Missing grabPay");
    }];
    
    [self waitForExpectationsWithTimeout:STPTestingNetworkRequestTimeout handler:nil];
}

@end

File: Stripe/StripeiOSTests/STPPaymentMethodGrabPayParamsTest.swift
//  Converted to Swift 5.8.1 by Swiftify v5.8.28463 - https://swiftify.com/
//
//  STPPaymentMethodGrabPayParamsTest.m
//  StripeiOS Tests
//
//  Created by Yuki Tokuhiro on 7/21/20.
//  Copyright © 2020 Stripe, Inc. All rights reserved.
//

import StripeCore
import StripeCoreTestUtils
import StripePaymentsTestUtils

class STPPaymentMethodGrabPayParamsTest: STPNetworkStubbingTestCase {
    func testCreateGrabPayPaymentMethod() {
        let client = STPAPIClient(publishableKey: STPTestingSGPublishableKey)
        let grabPayParams = STPPaymentMethodGrabPayParams()

        let billingDetails = STPPaymentMethodBillingDetails()
        billingDetails.name = "Jenny Rosen"

        let params = STPPaymentMethodParams(
            grabPay: grabPayParams,
            billingDetails: billingDetails,
            metadata: [
                "test_key": "test_value"
            ])

        let expectation = self.expectation(description: "Payment Method GrabPay create")

        client.createPaymentMethod(
            with: params) { paymentMethod, error in
            expectation.fulfill()

            XCTAssertNil(error, "Unexpected error creating GrabPay PaymentMethod")
            XCTAssertNotNil(paymentMethod, "Failed to create GrabPay PaymentMethod")
            XCTAssertNotNil(paymentMethod?.stripeId, "Missing stripeId")
            XCTAssertNotNil(paymentMethod?.created, "Missing created")
            XCTAssertFalse(paymentMethod!.liveMode, "Incorrect livemode")
                XCTAssertEqual(paymentMethod?.type, .grabPay, "Incorrect PaymentMethod type")

            // Billing Details
            XCTAssertEqual(paymentMethod?.billingDetails!.name, "Jenny Rosen")

            // GrabPay Details
            XCTAssertNotNil(paymentMethod?.grabPay, "Missing grabPay")
        }

        waitForExpectations(timeout: STPTestingNetworkRequestTimeout, handler: nil)
    }
}
